File: csharp/TetrisGameBase/logic/TetrisGame.cs
using hu.klenium.tetris.logic.board;
using hu.klenium.tetris.logic.tetromino;
using hu.klenium.tetris.util;

namespace hu.klenium.tetris.logic
{
    public class TetrisGame
    {
        protected bool isRunning;
        protected readonly Board board;
        protected Tetromino fallingTetromino = null;
        private readonly PeriodicTask gravity;

        public delegate void TetrominoStateChange(Tetromino tetromino);
        public delegate void BoardStateChange(Board board);
        public delegate void GameOverEvent();
        public event TetrominoStateChange OnTetrominoStateChanged;
        public event BoardStateChange OnBoardStateChanged;
        public event GameOverEvent OnGameOver;

        public TetrisGame(Dimension size, int fallingSpeed)
        {
            board = new Board(size);
            gravity = new PeriodicTask(() => HandleCommand(Command.FALL), fallingSpeed);
        }

        public void Start()
        {
            isRunning = GenerateNextTetromino();
            if (isRunning)
            {
                gravity.Start();
                OnTetrominoStateChanged?.Invoke(fallingTetromino);
                OnBoardStateChanged?.Invoke(board);
            }
        }
        private void Stop()
        {
            isRunning = false;
            gravity.Stop();
            OnGameOver?.Invoke();
        }

        public void HandleCommand(Command command)
        {
            if (!isRunning)
                return;
            bool stateChanged = false;
            switch (command)
            {
                case Command.ROTATE:
                    stateChanged = RotatetTetromino();
                    break;
                case Command.MOVE_LEFT:
                    stateChanged = MoveTetrominoLeft();
                    break;
                case Command.MOVE_RIGHT:
                    stateChanged = MoveTetrominoRight();
                    break;
                case Command.MOVE_DOWN:
                    gravity.Reset();
                    stateChanged = MoveTetrominoDown();
                    break;
                case Command.FALL:
                    stateChanged = MoveTetrominoDown();
                    break;
                case Command.DROP:
                    bool lastMovedDown;
                    do
                    {
                        lastMovedDown = MoveTetrominoDown();
                        stateChanged |= lastMovedDown;
                    } while (lastMovedDown);
                    break;
            }
            if (stateChanged && isRunning)
                OnTetrominoStateChanged?.Invoke(fallingTetromino);
        }

        private void TetrominoLanded()
        {
            board.AddTetromino(fallingTetromino);
            board.RemoveFullRows(fallingTetromino.Position.y, fallingTetromino.BoundingBox.height);
            OnBoardStateChanged?.Invoke(board);
            bool tetrominoAdded = GenerateNextTetromino();
            if (tetrominoAdded)
            {
                gravity.Reset();
                OnTetrominoStateChanged?.Invoke(fallingTetromino);
            }
            else
                Stop();
        }
        private bool GenerateNextTetromino()
        {
            int type = GetNextTetrominoType();
            fallingTetromino = new Tetromino(type, board);
            return fallingTetromino.MoveToInitialPos();
        }
        protected virtual int GetNextTetrominoType()
        {
            return Random.FromRange(0, 6);
        }
        private bool MoveTetrominoLeft()
        {
            return fallingTetromino.MoveLeft();
        }
        private bool MoveTetrominoDown()
        {
            bool moved = fallingTetromino.MoveDown();
            if (!moved)
                TetrominoLanded();
            return moved;
        }
        private bool MoveTetrominoRight()
        {
            return fallingTetromino.MoveRight();
        }
        private bool RotatetTetromino()
        {
            return fallingTetromino.RotateRight();
        }
    }
}

File: lua/src/tetrisgame.lua
require 'math'
local Board = require 'board'
local Tetromino = require 'tetromino'

math.randomseed(os.time())

local function TetrisGame(size)
	local self = {
		is_running = false,
		falling_tetromino = nil,
		board = Board(size)
	}
	
	local function stop()
        self.is_running = false
	end
    local function generate_next_tetromino()
        shape_type = self.get_next_tetromino_type()
        self.falling_tetromino = Tetromino(shape_type, self.board)
        return self.falling_tetromino.move_to_initial_pos()
	end
    local function tetromino_landed()
        self.board.add_tetromino(self.falling_tetromino)
        y = self.falling_tetromino.position.y
        height = self.falling_tetromino.bounding_box.height
        self.board.remove_full_rows(y, height)
        added = generate_next_tetromino()
        if not added then
            stop()
		end
	end
    local function rotate_tetromino()
        return self.falling_tetromino.rotate_right()
	end
    local function move_left_tetromino()
		print('tetromino move left')
        return self.falling_tetromino.move_left()
	end
    local function move_down_tetromino()
        local moved = self.falling_tetromino.move_down()
        if not moved then
            tetromino_landed()
		end
        return moved
	end
    local function move_right_tetromino()
        return self.falling_tetromino.move_right()
	end
	
    function self.start()
        self.is_running = generate_next_tetromino()
	end
    function self.handle_command(command)
        if not self.is_running then
            return
		end
		local actions = {
			    ROTATE = rotate_tetromino,
			 MOVE_LEFT = move_left_tetromino,
			MOVE_RIGHT = move_right_tetromino,
			 MOVE_DOWN = move_down_tetromino,
			      FALL = move_down_tetromino,
			      DROP = function()
						       while move_down_tetromino() do end
						   end
		}
		local action = actions[command]
		if action ~= nil then
			action()
		end
	end
    function self.get_next_tetromino_type()
        return math.random(1, 7)
	end

	return self
end

return TetrisGame